File: ml/train_model.py
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import joblib
import json
from prepare_training_data import extract_features

def train_email_classifier():
    """Train and save email classification model"""
    
    # Load training data
    df = pd.read_csv('ml/training_data.csv')
    
    # Separate features and labels
    feature_columns = [col for col in df.columns if col != 'label']
    X = df[feature_columns].values
    y = df['label'].values
    
    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )
    
    # Scale features
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    # Try different models
    models = {
        'RandomForest': RandomForestClassifier(
            n_estimators=100,
            max_depth=10,
            min_samples_split=2,
            min_samples_leaf=1,
            random_state=42
        ),
        'GradientBoosting': GradientBoostingClassifier(
            n_estimators=100,
            learning_rate=0.1,
            max_depth=5,
            random_state=42
        ),
        'LogisticRegression': LogisticRegression(
            max_iter=1000,
            random_state=42
        ),
        'SVM': SVC(
            kernel='rbf',
            probability=True,
            random_state=42
        )
    }
    
    best_model = None
    best_score = 0
    best_model_name = None
    
    print("Training and evaluating models...")
    print("-" * 50)
    
    for name, model in models.items():
        # Train model
        if name in ['LogisticRegression', 'SVM']:
            model.fit(X_train_scaled, y_train)
            y_pred = model.predict(X_test_scaled)
            train_data = X_train_scaled
        else:
            model.fit(X_train, y_train)
            y_pred = model.predict(X_test)
            train_data = X_train
        
        # Evaluate
        accuracy = accuracy_score(y_test, y_pred)
        
        # Cross-validation
        cv_scores = cross_val_score(model, train_data, y_train, cv=5)
        
        print(f"\n{name}:")
        print(f"Test Accuracy: {accuracy:.3f}")
        print(f"CV Score: {cv_scores.mean():.3f} (+/- {cv_scores.std() * 2:.3f})")
        print(f"Confusion Matrix:\n{confusion_matrix(y_test, y_pred)}")
        
        if accuracy > best_score:
            best_score = accuracy
            best_model = model
            best_model_name = name
    
    print("\n" + "=" * 50)
    print(f"Best Model: {best_model_name} with accuracy {best_score:.3f}")
    
    # Train final model on all data
    if best_model_name in ['LogisticRegression', 'SVM']:
        X_all_scaled = scaler.fit_transform(X)
        best_model.fit(X_all_scaled, y)
        # Save scaler
        joblib.dump(scaler, 'ml/scaler.pkl')
    else:
        best_model.fit(X, y)
    
    # Save model and metadata
    joblib.dump(best_model, 'ml/email_classifier.pkl')
    
    metadata = {
        'model_type': best_model_name,
        'accuracy': float(best_score),
        'feature_columns': feature_columns,
        'classes': ['CARGO', 'VESSEL'],
        'needs_scaling': best_model_name in ['LogisticRegression', 'SVM']
    }
    
    with open('ml/model_metadata.json', 'w') as f:
        json.dump(metadata, f, indent=2)
    
    print(f"\nModel saved to ml/email_classifier.pkl")
    print(f"Metadata saved to ml/model_metadata.json")
    
    # Test with sample emails
    print("\n" + "=" * 50)
    print("Testing with sample emails:")
    
    test_emails = [
        {
            "subject": "Steel coils 5000 MT ready for shipment",
            "body": "We have steel coils ready at Shanghai port. Need vessel for Japan",
            "sender": "steel@export.cn"
        },
        {
            "subject": "MV Star Eagle open position",
            "body": "Bulk carrier 55,000 DWT open Singapore next week",
            "sender": "ops@bulkship.sg"
        }
    ]
    
    for email in test_emails:
        features = extract_features(email)
        features_df = pd.DataFrame([features])[feature_columns]
        
        if metadata['needs_scaling']:
            features_array = scaler.transform(features_df.values)
        else:
            features_array = features_df.values
        
        prediction = best_model.predict(features_array)[0]
        probability = best_model.predict_proba(features_array)[0]
        
        print(f"\nEmail: {email['subject'][:50]}...")
        print(f"Prediction: {prediction}")
        print(f"Confidence: CARGO={probability[0]:.2f}, VESSEL={probability[1]:.2f}")

if __name__ == "__main__":
    # First prepare the data
    import prepare_training_data
    df = prepare_training_data.prepare_dataset()
    df.to_csv('ml/training_data.csv', index=False)
    
    # Train the model
    train_email_classifier()

File: ml/train_model.js
const fs = require('fs');
const path = require('path');

// Training data
const trainingData = [
    // CARGO emails
    {
        subject: "New Cargo Inquiry - 500 MT Steel Pipes",
        body: "We have a cargo of 500 metric tons of steel pipes ready for shipment from Shanghai to Rotterdam. Looking for suitable vessel.",
        sender: "cargo@logistics.com",
        type: "CARGO"
    },
    {
        subject: "Urgent: Container booking needed",
        body: "Need to book 20 TEU containers for electronics shipment. Origin: Shenzhen, Destination: Hamburg",
        sender: "shipper@export.cn",
        type: "CARGO"
    },
    {
        subject: "Bulk cargo - Grain shipment",
        body: "50,000 MT wheat cargo available ex Brazil ports. Looking for Panamax vessel for voyage to Mediterranean",
        sender: "graintrader@agri.br",
        type: "CARGO"
    },
    {
        subject: "Project cargo inquiry",
        body: "Heavy lift cargo consisting of turbines and generators. Total weight 200 MT. Need specialized vessel with cranes",
        sender: "project@energy.de",
        type: "CARGO"
    },
    {
        subject: "Chemical tanker cargo",
        body: "5000 MT caustic soda solution for shipment. IMO Class 8. Need chemical tanker with stainless steel tanks",
        sender: "chemical@trade.com",
        type: "CARGO"
    },
    {
        subject: "Coal shipment inquiry",
        body: "Steam coal cargo 75,000 MT available at Indonesian ports. Looking for Capesize vessel",
        sender: "coal@mining.id",
        type: "CARGO"
    },
    {
        subject: "Reefer cargo - Frozen food",
        body: "1000 MT frozen chicken cargo. Temperature requirement -18C. Need reefer vessel or containers",
        sender: "food@export.th",
        type: "CARGO"
    },
    {
        subject: "Iron ore cargo tender",
        body: "Iron ore fines 170,000 MT from Australia to China. Capesize vessel required",
        sender: "ironore@resources.au",
        type: "CARGO"
    },
    
    // VESSEL emails
    {
        subject: "MV Pacific Dream - Open for cargo",
        body: "Vessel MV Pacific Dream, Handymax bulk carrier, DWT 58,000 MT, open at Singapore from next week. Ready for grain/coal cargoes",
        sender: "operations@shipping.sg",
        type: "VESSEL"
    },
    {
        subject: "Container vessel available",
        body: "Container vessel 8,500 TEU capacity available for charter. Current position Mediterranean. Can accommodate reefer containers",
        sender: "chartering@maritime.gr",
        type: "VESSEL"
    },
    {
        subject: "Tanker vessel for spot charter",
        body: "Product tanker, 50,000 DWT, double hull, available for spot charter. All certificates valid. Position: Persian Gulf",
        sender: "tanker@fleet.ae",
        type: "VESSEL"
    },
    {
        subject: "Bulk carrier open position",
        body: "Panamax bulk carrier, 75,000 DWT, 5 holds/5 hatches, geared with 4x30T cranes. Open Japan next month",
        sender: "bulkops@vessel.jp",
        type: "VESSEL"
    },
    {
        subject: "Chemical tanker available",
        body: "Chemical tanker 20,000 DWT, stainless steel tanks, IMO II/III. Valid certificates. Open Houston area",
        sender: "chemical@tankers.us",
        type: "VESSEL"
    },
    {
        subject: "VLCC for time charter",
        body: "VLCC 300,000 DWT available for 1-year time charter. Double hull, 15 years old. All surveys passed",
        sender: "vlcc@supertanker.no",
        type: "VESSEL"
    },
    {
        subject: "General cargo vessel",
        body: "General cargo vessel 12,000 DWT with box-shaped holds. Good for project cargo. Open Mediterranean",
        sender: "general@cargo.it",
        type: "VESSEL"
    }
];

// Keyword lists for classification
const cargoKeywords = [
    'cargo', 'shipment', 'loading', 'discharge', 'commodity', 'mt', 'metric tons',
    'teu', 'container', 'bulk', 'breakbulk', 'project cargo', 'reefer',
    'grain', 'coal', 'iron ore', 'steel', 'chemical', 'oil', 'lng', 'cement',
    'timber', 'logs', 'rice', 'wheat', 'sugar', 'fertilizer', 'bauxite',
    'need vessel', 'looking for vessel', 'require vessel', 'booking',
    'freight rate', 'laycan', 'load port', 'discharge port', 'destination'
];

const vesselKeywords = [
    'vessel', 'ship', 'mv', 'm/v', 'dwt', 'draft', 'loa', 'beam', 'open',
    'available', 'position', 'charter', 'hire', 'tc', 'time charter',
    'voyage charter', 'spot', 'panamax', 'capesize', 'handymax', 'handysize',
    'bulk carrier', 'container vessel', 'general cargo', 'multipurpose',
    'tanker', 'chemical tanker', 'product tanker', 'lng carrier',
    'crane', 'gear', 'geared', 'gearless', 'holds', 'hatches',
    'ice class', 'double hull', 'certificates', 'class', 'flag'
];

// Feature extraction function
function extractFeatures(email) {
    const text = `${email.subject || ''} ${email.body || ''}`.toLowerCase();
    const sender = (email.sender || '').toLowerCase();
    
    // Count keyword occurrences
    const cargoScore = cargoKeywords.reduce((score, keyword) => {
        return score + (text.includes(keyword.toLowerCase()) ? 1 : 0);
    }, 0);
    
    const vesselScore = vesselKeywords.reduce((score, keyword) => {
        return score + (text.includes(keyword.toLowerCase()) ? 1 : 0);
    }, 0);
    
    // Check for tonnage patterns (e.g., "500 MT", "50,000 DWT")
    const hasTonnage = /\d+[\s,]*(?:mt|tons?|dwt|teu|cbm)/i.test(text);
    
    // Check for vessel name patterns (e.g., "MV Pacific Dream")
    const hasVesselName = /m\/?v\s+[\w\s]+/i.test(text);
    
    // Check sender domain
    const isShippingDomain = ['shipping', 'maritime', 'vessel', 'fleet', 'tanker', 'bulk']
        .some(domain => sender.includes(domain));
    const isCargoomain = ['cargo', 'logistics', 'export', 'import', 'trade', 'commodity']
        .some(domain => sender.includes(domain));
    
    return {
        cargoScore,
        vesselScore,
        hasTonnage,
        hasVesselName,
        isShippingDomain,
        isCargoomain,
        textLength: text.length
    };
}

// Simple rule-based classifier
class EmailClassifier {
    constructor() {
        this.threshold = 0.5;
    }
    
    train(data) {
        // Calculate average features for each class
        const cargoEmails = data.filter(e => e.type === 'CARGO');
        const vesselEmails = data.filter(e => e.type === 'VESSEL');
        
        this.cargoProfile = this.calculateProfile(cargoEmails);
        this.vesselProfile = this.calculateProfile(vesselEmails);
        
        console.log('Model trained successfully');
        console.log('Cargo profile:', this.cargoProfile);
        console.log('Vessel profile:', this.vesselProfile);
    }
    
    calculateProfile(emails) {
        const features = emails.map(e => extractFeatures(e));
        const profile = {};
        
        if (features.length === 0) return profile;
        
        // Calculate average for each feature
        profile.avgCargoScore = features.reduce((sum, f) => sum + f.cargoScore, 0) / features.length;
        profile.avgVesselScore = features.reduce((sum, f) => sum + f.vesselScore, 0) / features.length;
        profile.hasTonnageRate = features.filter(f => f.hasTonnage).length / features.length;
        profile.hasVesselNameRate = features.filter(f => f.hasVesselName).length / features.length;
        
        return profile;
    }
    
    predict(email) {
        const features = extractFeatures(email);
        
        // Calculate similarity scores
        let cargoSimilarity = 0;
        let vesselSimilarity = 0;
        
        // Weight cargo keywords more for cargo classification
        if (features.cargoScore > features.vesselScore) {
            cargoSimilarity += 0.6;
        } else if (features.vesselScore > features.cargoScore) {
            vesselSimilarity += 0.6;
        }
        
        // Check specific patterns
        if (features.hasVesselName) {
            vesselSimilarity += 0.3;
        }
        
        if (features.isCargoomain) {
            cargoSimilarity += 0.2;
        }
        
        if (features.isShippingDomain) {
            vesselSimilarity += 0.2;
        }
        
        // Normalize scores
        const totalScore = cargoSimilarity + vesselSimilarity;
        if (totalScore > 0) {
            cargoSimilarity /= totalScore;
            vesselSimilarity /= totalScore;
        }
        
        return {
            type: cargoSimilarity > vesselSimilarity ? 'CARGO' : 'VESSEL',
            confidence: {
                CARGO: cargoSimilarity,
                VESSEL: vesselSimilarity
            },
            features
        };
    }
    
    save(filepath) {
        const model = {
            cargoProfile: this.cargoProfile,
            vesselProfile: this.vesselProfile,
            threshold: this.threshold,
            version: '1.0.0'
        };
        
        fs.writeFileSync(filepath, JSON.stringify(model, null, 2));
        console.log(`Model saved to ${filepath}`);
    }
    
    load(filepath) {
        const model = JSON.parse(fs.readFileSync(filepath, 'utf8'));
        this.cargoProfile = model.cargoProfile;
        this.vesselProfile = model.vesselProfile;
        this.threshold = model.threshold;
        console.log(`Model loaded from ${filepath}`);
    }
}

// Train and save the model
function trainAndSaveModel() {
    const classifier = new EmailClassifier();
    
    // Train the model
    classifier.train(trainingData);
    
    // Test the model
    console.log('\nTesting the model:');
    console.log('=' .repeat(50));
    
    const testEmails = [
        {
            subject: "Steel coils 5000 MT ready for shipment",
            body: "We have steel coils ready at Shanghai port. Need vessel for Japan",
            sender: "steel@export.cn"
        },
        {
            subject: "MV Star Eagle open position",
            body: "Bulk carrier 55,000 DWT open Singapore next week",
            sender: "ops@bulkship.sg"
        }
    ];
    
    testEmails.forEach((email, idx) => {
        const prediction = classifier.predict(email);
        console.log(`\nTest ${idx + 1}: ${email.subject}`);
        console.log(`Prediction: ${prediction.type}`);
        console.log(`Confidence: CARGO=${prediction.confidence.CARGO.toFixed(2)}, VESSEL=${prediction.confidence.VESSEL.toFixed(2)}`);
    });
    
    // Save the model
    const modelPath = path.join(__dirname, 'email_classifier_model.json');
    classifier.save(modelPath);
    
    // Save training data for reference
    const dataPath = path.join(__dirname, 'training_data.json');
    fs.writeFileSync(dataPath, JSON.stringify(trainingData, null, 2));
    console.log(`Training data saved to ${dataPath}`);
    
    return classifier;
}

// Export for use in other modules
module.exports = {
    EmailClassifier,
    extractFeatures,
    trainAndSaveModel
};

// Run if called directly
if (require.main === module) {
    trainAndSaveModel();
}